Web Inspector: Sources: Sort resources without the file extension first

jdatapple opened the pull request:
#### 5060974c125e844e2b23d23d10e895661cfdd002
<pre>
Web Inspector: Sources: Sort resources without the file extension first
<a href="https://bugs.webkit.org/show_bug.cgi?id=220437">https://bugs.webkit.org/show_bug.cgi?id=220437</a>
rdar://72905353

Reviewed by Patrick Angle.

Sorting without the file extension first avoids comparing the dot (.) to
longer filenames with the same root string.
(e.g. styles.css and styles-dark.css)

* Source/WebInspectorUI/UserInterface/Views/ResourceTreeElement.js:
(WI.ResourceTreeElement.compareResourceTreeElements.filenameWithoutExtention):
(WI.ResourceTreeElement.compareResourceTreeElements):

Canonical link: <a href="https://commits.webkit.org/268557@main">https://commits.webkit.org/268557@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e5b010565f5ac43273dae801c75a7131d888604d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20032 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20470 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21091 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21936 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/18701 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23713 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20628 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20195 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20252 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20181 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/17413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22789 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/17351 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18210 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/24471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18428 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18386 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/22461 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [❌ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18974 "Failed to checkout and rebase branch from PR 18298") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/16100 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18169 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4797 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22514 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/18800 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ResourceTreeElement.js:
```diff
@@ -84,7 +84,19 @@ WI.ResourceTreeElement = class ResourceTreeElement extends WI.SourceCodeTreeElem
 
-        // Compare by title when the subtitles are the same.
+        // Compare by title without file extensions when the subtitles are the same.
+        function filenameWithoutExtention (resourceTitle) {
```
Style: No space between function name and opening paren.

patrickangle replied:
Nit: s/extention/extension here and throughout below.

jdatapple replied:
Addressed in 673a204c276c6ecb5d1c3dc8d0df42c54c1b0b67.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ResourceTreeElement.js:
```diff
@@ -84,7 +84,19 @@ WI.ResourceTreeElement = class ResourceTreeElement extends WI.SourceCodeTreeElem
+        let titleB = filenameWithoutExtention(b.mainTitle);
+        comparisonResult = titleA.extendedLocaleCompare(titleB);
+        if (comparisonResult !== 0)
+            return titleA.extendedLocaleCompare(titleB)
```
Nit: can just `return comparisonResult;`

jdatapple replied:
Addressed in 673a204c276c6ecb5d1c3dc8d0df42c54c1b0b67.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ResourceTreeElement.js:
```diff
@@ -84,7 +84,19 @@ WI.ResourceTreeElement = class ResourceTreeElement extends WI.SourceCodeTreeElem
-        // Compare by title when the subtitles are the same.
+        // Compare by title without file extensions when the subtitles are the same.
+        function filenameWithoutExtention (resourceTitle) {
+            let extentionPosition = resourceTitle.lastIndexOf("." + WI.fileExtensionForFilename(resourceTitle));
```
Interestingly the problem you appear to be trying to avoid by including the WI.fileExtensionForFilename in the lastIndexOf check, which is non-extension endings, doesn't actually work here. `WI.fileExtensionForFilename` will give you the string following the last index of a `.` in the name, unless the `.` is at the end of the string. Given that, it seems like for your purpose here we could just check for the last index of a `.` and get the same result.

jdatapple replied:
Addressed in 673a204c276c6ecb5d1c3dc8d0df42c54c1b0b67.

patrickangle reviewed (commented):
Seems like a good adjustment IMO.

patrickangle reviewed (approved):
`Extension` is still spelled wrong in the inner function name, but other than that LGTM

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ResourceTreeElement.js:
```diff
@@ -84,7 +84,18 @@ WI.ResourceTreeElement = class ResourceTreeElement extends WI.SourceCodeTreeElem
-        // Compare by title when the subtitles are the same.
+        // Compare by title without file extensions when the subtitles are the same.
+        function filenameWithoutExtension(resourceTitle) {
+            return resourceTitle.substring(0, resourceTitle.lastIndexOf("."));
```
i actually wonder if maybe we should be sorting based on the _first_ `"."`

for example, id imagine `test.min.js` to be before `test-a.js`

if so, we also could maybe be a bit more efficient in how we compare our strings by using `split(".", 1)`
```
function splitFilename(resourceTitle) {
    let [name, extensions] = resourceTitle.split(".", 1);
    return {name, extensions: extensions || ""};
}
let splitA = splitFilename(a.mainTitle);
let splitB = splitFilename(b.mainTitle);
return splitA.name.extendedLocaleCompare(splitB.name) || splitA.extensions.extendedLocaleCompare(splitB.extensions);
```

not a big deal tho if you find that less clear to read

jdatapple replied:
This is a great idea! I implemented this in: https://github.com/WebKit/WebKit/pull/18350

jdatapple committed:
Web Inspector: Sources: Sort resources without the file extension first
https://bugs.webkit.org/show_bug.cgi?id=220437
rdar://72905353

Reviewed by Patrick Angle.

Sorting without the file extension first avoids comparing the dot (.) to
longer filenames with the same root string.
(e.g. styles.css and styles-dark.css)

* Source/WebInspectorUI/UserInterface/Views/ResourceTreeElement.js:
(WI.ResourceTreeElement.compareResourceTreeElements.filenameWithoutExtention):
(WI.ResourceTreeElement.compareResourceTreeElements):

Canonical link: https://commits.webkit.org/268557@main
